nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java

```diff
@@ -158,41 +173,51 @@
+            final boolean accountKeyIsSet = validationContext.getProperty(ACCOUNT_KEY).isSet();
+            final boolean sasTokenIsSet = validationContext.getProperty(SAS_TOKEN).isSet();
+
+            int credential_config_found = 0;
+            if (useManagedIdentity) credential_config_found++;
+            if (accountKeyIsSet) credential_config_found++;
+            if (sasTokenIsSet) credential_config_found++;
+
+            if (credential_config_found == 0) {
+                final String msg = String.format(
+                    "At least one of ['%s', '%s', '%s'] should be set",
+                    ACCOUNT_KEY.getDisplayName(),
+                    SAS_TOKEN.getDisplayName(),
+                    USE_MANAGED_IDENTITY.getDisplayName()
+                );
+                results.add(new ValidationResult.Builder().subject("Credentials config").valid(false).explanation(msg).build());
+            } else if (credential_config_found > 1) {
+                final String msg = String.format(
+                    "Only one of ['%s', '%s', '%s'] should be set",
+                    ACCOUNT_KEY.getDisplayName(),
+                    SAS_TOKEN.getDisplayName(),
+                    USE_MANAGED_IDENTITY.getDisplayName()
+                );
+                results.add(new ValidationResult.Builder().subject("Credentials config").valid(false).explanation(msg).build());
+            }
         }
+
         return results;
     }
 
     public static DataLakeServiceClient getStorageClient(PropertyContext context, FlowFile flowFile) {
-        final Map<String, String> attributes = flowFile != null ? flowFile.getAttributes() : Collections.emptyMap();
-        final String accountName = context.getProperty(ACCOUNT_NAME).evaluateAttributeExpressions(attributes).getValue();
-        final String accountKey = context.getProperty(ACCOUNT_KEY).evaluateAttributeExpressions(attributes).getValue();
-        final String sasToken = context.getProperty(SAS_TOKEN).evaluateAttributeExpressions(attributes).getValue();
-        final String endpointSuffix = context.getProperty(ENDPOINT_SUFFIX).evaluateAttributeExpressions(attributes).getValue();
+        AzureStorageCredentialsDetails storageCredentialsDetails = AzureStorageUtils.getStorageCredentialsDetails(context, flowFile);
+
+        final String accountName = storageCredentialsDetails.getStorageAccountName();
+        final String accountKey = storageCredentialsDetails.getAccountKey();
+        final String sasToken = storageCredentialsDetails.getSasToken();
+        final AccessToken accessToken = storageCredentialsDetails.getAccessToken();
+
+        // ControllerService - if set - may or may not have endpoint suffix defined
+        final String endpointSuffix = Optional
+            .ofNullable(storageCredentialsDetails.getStorageSuffix())
+            .orElse(context.getProperty(ENDPOINT_SUFFIX).getValue());
```

Comment:
The processor level property is not a fallback option. When the CS defined, the processor level properties should be null (see my comment on `validateCredentialProperties()`).